### ISSUE TITLE
YOLOv8-OBB learns 0° rotation

### ISSUE THREAD

**Issue by simoneangarano** [ultralytics/ultralytics#13081]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

When training YOLOv8-OBB on a custom dataset with oriented bounding boxes, the model learns 0° rotation for every prediction, resulting in standard bounding boxes. I guess that the training loss does not penalize the model for predicting wrong angles. Can you please check and tell me how to modify the training code to fix this issue?
Thanks.

### Environment

Ultralytics YOLOv8.2.19 🚀 Python-3.9.18 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24245MiB)
Setup complete ✅ (24 CPUs, 62.5 GB RAM, 249.4/1832.2 GB disk)

### Minimal Reproducible Example

```
from ultralytics import YOLO
# Load a model
model = YOLO("yolov8n-obb.yaml")  # build a new model from YAML
# Train the model
results = model.train(data="dataset.yaml", epochs=100, imgsz=640)
```

### Additional

I cannot share the dataset, but in the training log images the batch visualizations show correct rotated bounding boxes while predictions are not rotated.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by samin50** [comment]
How are you displaying your bounding boxes? The results object that you get as a return value for predict has several bounding box coordinate types, for example results[0].obb.xyxyxyxy for 4 pairs of xy coords for each corner.

**Comment by simoneangarano** [comment]
results[0].plot(). Even the visualizations auto-generated during training (val_batch0_pred.jpg) show bboxes that are only slightly rotated.

**Comment by glenn-jocher** [comment]
@simoneangarano hello! It sounds like the model might be struggling to learn the angle variations effectively. This could be due to a variety of factors including the diversity of the training data in terms of angles or the specific characteristics of the loss function handling the angle predictions.

To better diagnose the issue, you could:
- Check the distribution of angles in your training dataset to ensure there's enough variation.
- Experiment with adjusting the loss weights for the angle component in the training configuration.

For visualizing the predictions with more clarity on rotations, you might consider manually plotting the bounding boxes using `results[0].obb.xyxyxyxy` to extract and visualize each box's corners. This can give you a clearer picture of how the model is predicting rotations.

Here's a quick example on how you can plot these manually for more detailed inspection:

```python
import matplotlib.pyplot as plt
import numpy as np

# Assuming 'result' is your prediction result for one image
boxes = result.obb.xyxyxyxy  # Get the oriented bounding box coordinates

fig, ax = plt.subplots(1)
ax.imshow(result.img)  # Plot the image

# Plot each OBB
for box in boxes:
    poly = plt.Polygon(np.array(box).reshape(-1, 2), closed=True, edgecolor='r', fill=None)
    ax.add_patch(poly)

plt.show()
```

This might help you visually confirm the model's performance on angle predictions more precisely.

**Comment by simoneangarano** [comment]
> * Experiment with adjusting the loss weights for the angle component in the training configuration.

How do I increase the weight for the angle component? I don't see any training hyperparameter that's specific for that.

**Comment by glenn-jocher** [comment]
@simoneangarano hello! In the current YOLOv8 implementation, direct hyperparameter adjustment for the angle component in the loss function isn't exposed via the training configuration. However, you can modify the source code where the loss is computed to manually increase the weight for the angle component.

Here’s a brief guide on how you might approach this:

1. Locate the loss computation section in the model's code (usually found in the model definition files).
2. Identify the part of the loss that corresponds to the angle prediction.
3. Increase the multiplier for the angle loss to give it more weight during training.

If you need specific guidance on which file or line to edit, I can help you further if you provide more details about your setup or the version of YOLOv8 you are using. 🚀

**Comment by simoneangarano** [comment]
Thanks!
Yes, I think I need more help figuring out where specifically the rotation loss is computed.
My setup is:
Ultralytics YOLOv8.2.19 🚀 Python-3.9.18 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24245MiB)
Setup complete ✅ (24 CPUs, 62.5 GB RAM, 249.4/1832.2 GB disk)
I'm using YOLOv8-OBB.

**Comment by simoneangarano** [comment]
I guess the angles were not being considered in the default loss, as manually adding an additional loss component ´ang_loss´ solves the issue. Now the model learns to predict rotated bounding boxes. Still, what is the best loss to be used here? I used MSE and it seems to be working, but maybe there is something better. Help?

This is the code I modified in ´ultralytics/utils/loss.py´. If you want to include it in the next release, I can help by creating a pull request.

``` python3
# Cls loss
# loss[1] = self.varifocal_loss(pred_scores, target_scores, target_labels) / target_scores_sum  # VFL way
loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE

# Bbox loss
if fg_mask.sum():
    target_bboxes[..., :4] /= stride_tensor
    loss[0], loss[2] = self.bbox_loss(
        pred_distri, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask
    )
else:
    loss[0] += (pred_angle * 0).sum()

angle_diff = target_bboxes[fg_mask][:,-1] - pred_bboxes[fg_mask][:,-1]
angle_loss = (angle_diff ** 2).mean()

loss[0] *= self.hyp.box  # box gain
loss[1] *= self.hyp.cls  # cls gain
loss[2] *= self.hyp.dfl  # dfl gain
loss[3] = angle_loss * self.hyp.ang

return loss.sum() * batch_size, loss.detach()  # loss(box, cls, dfl)

```

**Comment by glenn-jocher** [comment]
Hey there! 🚀 Great job on integrating the angle loss into the model! Using MSE for the angle loss is a solid choice as it emphasizes larger errors and is generally well-behaved during optimization. However, you might also consider experimenting with the Smooth L1 Loss, which is less sensitive to outliers than MSE. This can sometimes lead to better performance, especially in cases where the angle variation is large.

Here's how you could modify your code to use Smooth L1 Loss for the angle:

```python
angle_diff = target_bboxes[fg_mask][:,-1] - pred_bboxes[fg_mask][:,-1]
angle_loss = F.smooth_l1_loss(pred_bboxes[fg_mask][:,-1], target_bboxes[fg_mask][:,-1], reduction='mean')
```

This change might provide a more robust training process with respect to angle predictions. If you're seeing good results and want to contribute, a pull request would be fantastic! The community would definitely benefit from your insights and improvements. 😊

**Comment by simoneangarano** [comment]
Thanks for the help! I will.

**Comment by glenn-jocher** [comment]
@simoneangarano you're welcome! 😊 If you have any more questions or need further assistance, feel free to reach out. Looking forward to your pull request! 🚀

**Comment by Petros626** [comment]
@simoneangarano what are your results?

**Comment by simoneangarano** [comment]
> [@simoneangarano](https://github.com/simoneangarano) what are your results?

Hi, I'm sorry I didn't reply to your comment on GitHub! Unfortunately, I can't share the results as I used NDA-protected data. The difference was pretty big, though. Default angle loss was not working, and the model always learned 0° rotations. With my modification, the model correctly detected rotated objects. I don't know if newer versions of ultralytics already solved the problem. Hope this helps!

**Comment by Petros626** [comment]
I just wanted to be sure, that your approach worked fine. Can you give at least some details, how you improved the angle learning (loss function etc.). What exactly you used from the above code?

**Comment by simoneangarano** [comment]
All my code is in [this pull request](https://github.com/ultralytics/ultralytics/pull/13157). Just use the standard configuration for `yolov8n-obb.yaml`.

You will also notice that I added a new configuration parameter, `ang: 5.0`, as an angle loss gain in the configuration file.

**Comment by Petros626** [comment]
Thank you very much. @glenn-jocher could be worth a PR, when results are promising?

**Comment by glenn-jocher** [comment]
@Petros626 thank you for sharing your solution and submitting the pull request! We appreciate contributions like yours to improve YOLO-OBB training. Your approach of adding an angle loss component (MSE) with a configurable `ang` hyperparameter aligns with our goal of making the framework more flexible. Community members can review your implementation in [PR#13157](https://github.com/ultralytics/ultralytics/pull/13157) and test it with their datasets. This type of collaborative improvement is exactly what drives open-source innovation forward. 🚀
